rfcs/20200618-headers/README.md

```diff
@@ -0,0 +1,122 @@
+# Changes in the oneDNN header install locations and namespace
+
+## Introduction
+
+New headers structure and namespaces are needed for cross-component and oneAPI
+branding alignment. The request from the oneAPI program is to put oneDNN
+headers under the `oneapi` and also to provide a corresponding namespace. The
+changes will have to be reflected in the
+[spec](https://github.com/oneapi-src/oneAPI-spec/tree/master/source/elements/oneDNN).
+
+The main concern is how to keep backward compatibility with the existing code.
+The secondary concern is maintenance burden.
+
+## Proposal
+
+The following changes are to be implemented in the
+[`dev-v2`](https://github.com/oneapi-src/onednn/tree/dev-v2) branch but will
+eventually be part of master.
+
+### Namespace changes
+
+oneDNN should start providing two namespaces (C++ only): `dnnl::` and
+`oneapi::dnnl::`. Either can be an alias. To minimize the changes, the
+proposal is to alias the long one:
+
+```c++
+namespace dnnl { /* the usual stuff */ }
+
+namespace oneapi { namespace dnnl = ::dnnl; }
+```
+
+The new namespace must be reflected in the
+[spec](https://github.com/oneapi-src/oneAPI-spec/tree/master/source/elements/oneDNN).
+There is no recommendation on whether longer or shorted namespace should be
+used in the spec yet; this is outside of the scope of this RFC.
+
+### Installation paths changes
+
+oneDNN is free to decide on the header structure under `include/oneapi`.
+
+#### Option 1 -- recommended
+
+```
+include
+├── oneapi[/dnnl]
+│   ├── dnnl_types.h
+│   ├── dnnl_config.h
+│   ├── dnnl.hpp
+│   └── dnnl.h
+├── dnnl_types.h        (#include "oneapi[/dnnl]/dnnl_types.h")
+├── dnnl_config.h       (#include "oneapi[/dnnl]/dnnl_config.h")
+├── dnnl.hpp            (#include "oneapi[/dnnl]/dnnl.hpp"")
+└── dnnl.h              (#include "oneapi[/dnnl]/dnnl.h"")
+```
+
+The `/dnnl/` portion of the path may be present or not. The recommendation is
+*to include it* so that it is possible to extend dnnl library later in a
+straightforward manner.
+
+Additionally, the compatibility headers can complain like this:
+
+```c++
+#ifdef _MSC_VER
+#pragma message "Warning: this header is deprecated. Please include oneapi[/dnnl]/dnnl.hpp instead"
+#else
+#warning "This header is deprecated. Please include oneapi[/dnnl]/dnnl.hpp instead"
+#endif
+
+#include "oneapi[/dnnl]/dnnl.hpp"
+```
```

Comment:
we haven't done that for mkldnn.h* compat headers, I suggest we don't do that now as well...

Reply:
we have been bringing to much confusion :(

Reply:
Unrecommended.

Reply:
For framework like Tensorflow: for now, while TF 2.x is using OneDNN 1.5. Is there any change needs to be done for now, regarding integrating OneDNN to TF? The change needs to be done during OneDNN 2.x integration, is it correct?

Reply:
Deprecation warnings make sense if you plan to remove the headers and want to inform users in advance. If you prefer to keep the compat headers for long, warnings are excessive.

Reply:
In oneTBB we do use warnings for deprecated features and headers but also provide a macro to switch those off. The intent is to raise awareness, particularly if removal is planned within a year or so.